contracts/protocore/SelfProtocore.sol

```diff
@@ -70,4 +91,142 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
+     * @param _sourceKernelHash Hash of kernel at origin chain.
+     * @param _sourceDynasty Dynasty of origin chain.
+     * @param _sourceAccumulatedGas Accumulated gas at origin chain.
+     * @param _sourceCommitteeLock Committee lock at source.
+     */
+    function proposeLink(
+        bytes32 _parentVoteMessageHash,
+        bytes32 _targetBlockHash,
+        uint256 _targetBlockNumber,
+        bytes32 _sourceKernelHash,
+        bytes32 _sourceOriginObservation,
+        uint256 _sourceDynasty,
+        uint256 _sourceAccumulatedGas,
+        bytes32 _sourceCommitteeLock
+    )
+        external
+    {
+        require(
+            _parentVoteMessageHash != bytes32(0),
+            "Vote message hash of parent must not be 0."
+        );
+        require(
+            _targetBlockHash != bytes32(0),
+            "Target blockhash must not be 0."
+        );
+        require(
+            _sourceKernelHash != bytes32(0),
+            "Origin observation must not be 0."
+        );
+        require(
+            _sourceOriginObservation != bytes32(0),
+            "Origin observation must not be 0."
+        );
+        require(
+            _sourceCommitteeLock != bytes32(0),
+            "Source committee lock must not be 0."
+        );
+        require(
+            _sourceKernelHash == openKernelHash,
+            "Source kernel hash must match with already opened kernel hash."
+        );
+
+        bytes32 sourceTransitionHash = hashSourceTransition(
+            _sourceKernelHash,
+            _sourceOriginObservation,
+            _sourceDynasty,
+            _sourceAccumulatedGas,
+            _sourceCommitteeLock
+        );
+
+        Protocore.proposeLinkInternal(
+            _parentVoteMessageHash,
+            sourceTransitionHash,
+            _targetBlockHash,
+            _targetBlockNumber
+        );
+
+        emit LinkProposed(
+            _parentVoteMessageHash,
+            _targetBlockHash,
+            _targetBlockNumber,
+            _sourceOriginObservation,
+            _sourceKernelHash,
+            _sourceDynasty,
+            _sourceAccumulatedGas,
+            _sourceCommitteeLock
+        );
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice Takes the parameters of an source transition object and returns the
+     *         typed hash of it.
+     *
+     * @param _kernelHash Kernel hash.
+     * @param _originObservation Observation of the origin chain.
+     * @param _dynasty The dynasty number where the meta-block closes
+     *                 on the auxiliary chain.
+     * @param _accumulatedGas The total consumed gas on auxiliary within a
+     *                        meta-block.
+     * @param _committeeLock The committee lock that hashes the transaction
+     *                       root on the auxiliary chain.
+     * @return sourceTransitionHash_ The hash of source transition object.
+     */
+    function hashSourceTransition(
```

Comment:
```suggestion
    function hashTransition(
```
This should be same as `Core::hashTransition`.

Reply:
Done